browser/src/libs/code_intelligence/code_intelligence.tsx

```diff
@@ -477,213 +472,224 @@ export function handleCodeHost({
+            ])
+        }
+    }
+
+    /**
+     * Deletes a reference to a workspace root from a code editor.
+     *
+     * Will only cause `workspace.roots` to emit if the root
+     * with the given `uri` has no more references.
+     */
+    const deleteRootRef = (uri: string) => {
+        const currentRefCount = rootRefCounts.get(uri)
+        if (!currentRefCount) {
+            throw new Error('No preexisting root refs')
+        }
+        const updatedRefCount = currentRefCount - 1
+        if (updatedRefCount === 0) {
+            extensionsController.services.workspace.roots.next(
+                extensionsController.services.workspace.roots.value.filter(root => root.uri !== uri)
+            )
+        } else {
+            rootRefCounts.set(uri, updatedRefCount)
+        }
     }
-    /** Map from code view element to the state associated with it (to be updated or removed) */
-    const codeViewStates = new Map<Element, CodeViewState>()
 
     subscriptions.add(
         codeViews.subscribe(codeViewEvent => {
-            console.log(`Code view ${codeViewEvent.type}`)
-
-            // Handle added or removed view component, workspace root and subscriptions
-            if (codeViewEvent.type === 'added' && !codeViewStates.has(codeViewEvent.element)) {
-                const { element, fileInfo, getPositionAdjuster, getToolbarMount, toolbarButtonProps } = codeViewEvent
-                const uri = toURIWithPath(fileInfo)
-                const languageId = getModeFromPath(fileInfo.filePath)
-                const model = { uri, languageId, text: fileInfo.content }
+            console.log('Code view added')
+            codeViewEvent.subscriptions.add(() => console.log('Code view removed'))
+
+            const { element, fileInfo, getPositionAdjuster, getToolbarMount, toolbarButtonProps } = codeViewEvent
+            const uri = toURIWithPath(fileInfo)
+            const languageId = getModeFromPath(fileInfo.filePath)
+            const model = { uri, languageId, text: fileInfo.content }
+            // Only add the model if it doesn't exist
+            // (there may be several code views on the page pointing to the same model)
+            if (!extensionsController.services.model.hasModel(uri)) {
+                extensionsController.services.model.addModel(model)
+            }
+            const editorData: CodeEditorData = {
+                type: 'CodeEditor' as const,
+                resource: uri,
+                selections: codeViewEvent.getSelections ? codeViewEvent.getSelections(codeViewEvent.element) : [],
+                isActive: true,
+            }
+            const editorId = extensionsController.services.editor.addEditor(editorData)
+            const scope: PartialCodeEditor = {
+                ...editorData,
+                ...editorId,
+                model,
+            }
+            const rootURI = toRootURI(fileInfo)
+            addRootRef(rootURI, fileInfo.rev || '')
```

Comment:
Empty string seems wrong to me, should be `undefined` (which both `fileInfo.rev` and `WorkspaceRootWithMetadata.inputRev` allow
```suggestion
            addRootRef(rootURI, fileInfo.rev)
```